### PR TITLE
Add `enable_model_summary` flag

### LINKED ISSUE #9043: [RFC] Deprecate `weights_summary` off the Trainer constructor
## Proposed refactoring or deprecation

- Introduce a new ModelSummary callback, which calls `summarize`: https://github.com/PyTorchLightning/pytorch-lightning/blob/8a931732ae5135e3e55d9c7b7031d81837e5798a/pytorch_lightning/utilities/model_summary.py#L437-L439
- Deprecate `weights_summary` off the Trainer constructor

### Motivation

We are auditing the Lightning components and APIs to assess opportunities for improvements:
- https://github.com/PyTorchLightning/pytorch-lightning/issues/7740
- https://docs.google.com/document/d/1xHU7-iQSpp9KJTjI3As2EM0mfNHHr37WZYpDpwLkivA/edit#

This is a followup to https://github.com/PyTorchLightning/pytorch-lightning/issues/8478 and https://github.com/PyTorchLightning/pytorch-lightning/discussions/9006

Why do we want to remove this from the core trainer logic?
- We need a way for users to customize more of the inputs to the model summary over time without affecting the trainer API. Today, changes to the model summary API also require changes in the core trainer (e.g. the addition of `max_depth` ). This gives model summarization more room to grow without cascading changes elsewhere. 
- Users may want to configure this summarization for different points of execution. For instance, calling this at the start of `trainer.fit()`, `trainer.validate()`, `trainer.test()` or `trainer.predict()`. Right now, this is hardcoded to be run only during `fit()`. 
- Users may want to customize where they save the summary. Right now, it's printed to stdout, but this could also be useful to save to a file or upload to another service for tracking the run.
- The current implementation runs on global rank 0 only in order to avoid printing out multiple summary tables. However, running this on rank 0 will break for model parallel use cases that require communication across ranks. This can lead to subtle failures if `example_input_array` is set as a property on the LightningModule. For instance, a model wrapped with FSDP will break because parameters need to be all-gathered across layers across ranks.
- In case the LightningModule leverages PyTorch LazyModules, users may want to generate this summary only after the first batch is processed in order to get accurate parameter estimations. Estimates of parameter sizes with lazy modules would be misleading.

### Pitch

A callback in Lightning naturally fits this extension purpose. It generalizes well across lightning modules, has great flexibility for when it can be called, and allows users to customize the summarization logic (e.g. integrate other libraries more easily). 
- https://github.com/tyleryep/torchinfo
- https://github.com/facebookresearch/fvcore/blob/master/fvcore/nn/flop_count.py

- With this callback available, this logic can be removed from the core Trainer in order to be more pluggable: 
- https://github.com/PyTorchLightning/pytorch-lightning/blob/6604fc1344e1b8a459c45a5a2157aa7fc60d950d/pytorch_lightning/trainer/trainer.py#L1000-L1004 

AFAICT, this is the only piece of logic that runs in between `on_pretrain_routine_start/end` hooks. Would we still need these hooks if the summarization logic was removed from the trainer? Why doesn't this happen in `on_train_start` today? 
https://github.com/PyTorchLightning/pytorch-lightning/blob/8a931732ae5135e3e55d9c7b7031d81837e5798a/pytorch_lightning/trainer/trainer.py#L1103-L1113

### Additional context

The model summary is by default enabled right now. This is likely the core issue we have to resolve as to whether this is opt-in or opt-out: https://github.com/PyTorchLightning/pytorch-lightning/issues/8478#issuecomment-884533398

Seeking @edenafek @tchaton 's input on this

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Fixes #9043 
Adds `enable_model_summary` flag to match `enable_progress_bar` flag being added here: https://github.com/PyTorchLightning/pytorch-lightning/pull/9664/

This PR also deprecates the Trainer `weights_summary` flag and property which dictated the mode for display. Users are expected to customize this with the callback. This deprecation was required because we needed to support arbitrary module depths for summarization (int) while the current mode only supported a few fixed modes ("top" vs "full"). This customization is much easier with the ModelSummary callbacks @kaushikb11  introduced.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/callback_connector.py [hunk 1]
```diff
@@ -145,24 +146,43 @@ def _configure_checkpoint_callbacks(self, checkpoint_callback: bool) -> None:
...
+                " in v1.7. Please set `Trainer(enable_model_summary=False) instead."
+            )
+            return
+        # Prior default in the Trainer for `weights_summary` which we explicitly check here
+        # to preserve backwards compatibility
+        if weights_summary != "top":
+            rank_zero_deprecation(
+                f"Setting `Trainer(weights_summary={weights_summary})` is deprecated in v1.5 and will be removed"
+                " in v1.7. Please pass `pytorch_lightning.callbacks.model_summary.ModelSummary` with"
+                " `max_depth` directly to the Trainer's `callbacks` argument instead."
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/connectors/callback_connector.py:168]:
```suggestion
                f"Setting `Trainer(weights_summary={weights_summary})` is deprecated in v1.5 and will be removed "
                "in v1.7. Please pass `pytorch_lightning.callbacks.model_summary.ModelSummary` with "
                "`max_depth` directly to the Trainer's `callbacks` argument instead."
```
same here.

File: pytorch_lightning/trainer/connectors/callback_connector.py [hunk 2]
```diff
@@ -145,24 +146,43 @@ def _configure_checkpoint_callbacks(self, checkpoint_callback: bool) -> None:
...
+            return
         if any(isinstance(cb, ModelSummary) for cb in self.trainer.callbacks):
             return
-        if weights_summary is not None:
+        if weights_summary is None:
+            rank_zero_deprecation(
+                "Setting `Trainer(weights_summary=None)` is deprecated in v1.5 and will be removed"
+                " in v1.7. Please set `Trainer(enable_model_summary=False) instead."
+            )
+            return
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/connectors/callback_connector.py:161]: isn't the warning below enough? you can do a check below if it's None and return from there itself.

File: pytorch_lightning/trainer/connectors/callback_connector.py [hunk 3]
```diff
@@ -145,24 +146,43 @@ def _configure_checkpoint_callbacks(self, checkpoint_callback: bool) -> None:
         if not self._trainer_has_checkpoint_callbacks() and checkpoint_callback is True:
             self.trainer.callbacks.append(ModelCheckpoint())
 
-    def _configure_model_summary_callback(self, weights_summary: Optional[str] = None) -> None:
+    def _configure_model_summary_callback(
+        self, enable_model_summary: bool, weights_summary: Optional[str] = None
+    ) -> None:
+        if not enable_model_summary:
+            return
         if any(isinstance(cb, ModelSummary) for cb in self.trainer.callbacks):
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/connectors/callback_connector.py:154]: what if there is a callback and weights_summary is not None? shouldn't we raise an exception for that case?

Comment (ananthsub) [reply pytorch_lightning/trainer/connectors/callback_connector.py:154]: I don't see a way we can do this now, as the existing default is `weights_summary="top"`
If we change the default value to `weights_summary=None`, then it's a breaking change, as users today use `weights_summary=None` to configure whether this is displayed or not. After this PR, they'd be forced to use `enable_model_summary=False` since there'd be no way to distinguish precedence/user input vs defaults between `enable_model_summary=True` and `weights_summary=None` 

So if we want to preserve the existing disable path, we can't change the default for `weights_summary`. 
On the other hand, if someone is using a custom model summary callback, they need all of `enable_model_summary=True`, `weights_summary != None` and the custom callback extending `ModelSummary`. and raising an exception here would block that use case entirely.

After `weights_summary` is deprecated, we can definitely strengthen this check

File: pytorch_lightning/trainer/connectors/callback_connector.py (hunk 2 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/connectors/callback_connector.py:161]: I wanted to use a different warning message as `weights_summary=None` -> `enable_model_summary=False` whereas `weights_summary is not None` -> `callbacks=[ModelSummary(...)]`

File: pytorch_lightning/trainer/connectors/callback_connector.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/connectors/callback_connector.py:168]: in this file it looks like the spaces are all prepended to the following line besides the gradient accumulation error message

Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/callback_connector.py:189]: all the new ones are actually following this one (spaces appended at the end)

File: docs/source/common/trainer.rst [hunk 4]
```diff
@@ -1611,6 +1611,23 @@ Options: 'full', 'top', None.
     # don't print a summary
     trainer = Trainer(weights_summary=None)
 
+.. warning:: `weights_summary` is deprecated in v1.5 and will be removed in v1.7. Please pass :class:`~pytorch_lightning.callbacks.model_summary.ModelSummary`
```
Comment (rohitgr7) [inline docs/source/common/trainer.rst:1614]: should we put this at the top of this section?

File: docs/source/common/trainer.rst [hunk 5]
```diff
@@ -1611,6 +1611,23 @@ Options: 'full', 'top', None.
...
+
+Whether to enable or disable the model summarization. Defaults to True.
+
+.. testcode::
+
+    # default used by the Trainer
+    trainer = Trainer(enable_model_summary=True)
+
+    # disable summarization
+    trainer = Trainer(enable_model_summary=False)
```
Comment (rohitgr7) [inline docs/source/common/trainer.rst:1630]: shall we have a small example with `ModelSummary` callback here?

**Review by rohitgr7 — APPROVED** [review]
LG!. Just some minor comments.

File: CHANGELOG.md [hunk 6]
```diff
@@ -296,6 +296,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Deprecated Accelerator collective API `barrier`, `broadcast`, and `all_gather`, call `TrainingTypePlugin` collective API directly ([#9677](https://github.com/PyTorchLightning/pytorch-lightning/pull/9677))
 
 
+- Deprecated passing `weights_summary` to the `Trainer` constructor in favor of adding the `ModelSummary` callback with `max_depth` directly to the list of callbacks ([#9616](https://github.com/PyTorchLightning/pytorch-lightning/pull/9616))
+
+
```
Comment (kaushikb11) [inline CHANGELOG.md:301]: Would need to update the Changelog about adding the `enable_model_summary` flag

**Comment by ananthsub** [comment]
> Do we need the `enable_` prefix? I think `model_summary` is straightforward as it is a boolean flag.

This also came up in these places:
https://github.com/PyTorchLightning/pytorch-lightning/pull/9616#issuecomment-924387620
https://github.com/PyTorchLightning/pytorch-lightning/pull/9664#issuecomment-927314218
https://github.com/PyTorchLightning/pytorch-lightning/pull/9754#issuecomment-930257778

Options so far:

1. Use enable_ as a prefix
- Pros:
  - consistent naming sceheme: users can scan through the trainer flags and see what features are enabled by default by the Trainer
  - `enable_` as a a prefix indicates these are booleans
- Cons:
  - More verbose
2. Directly use the name. e.g. `model_summary`, `progress_bar`, `checkpointing`
- Pros:
  - Less verbose
- Cons:
  - May have confusion with the callback implementation (e.g. `model_summary` as a flag: does it take the `ModelSummary` callback? or a boolean? would `trainer.model_summary` be the callback instance, or the boolean flag passed into the trainer?
3. Other options?

@PyTorchLightning/core-contributors

**Comment by daniellepintz** [comment]
I am strongly in favor of option 1. Another reason is for consistency, i.e. consider these Trainer arguments:
* `profiler` of type `BaseProfiler` 
* `accelerator` of type `Accelerator`
* `logger` of type `LightningLoggerBase`

if I were reading these arguments and then saw one named `model_summary` I would assume it to be an instance of a class as well. 

IMO boolean arguments should be named with a verb as a prefix, such as `enable_progress_bar`, `reload_dataloaders_every_epoch`, `terminate_on_nan`, `move_metrics_to_cpu`

**Comment by daniellepintz** [comment]
Question: under what circumstances would users want to disable the model summary? Is it critical we have this flag?

File: docs/source/common/debugging.rst [hunk 7]
```diff
@@ -95,11 +95,14 @@ Print a summary of your LightningModule
 ---------------------------------------
 Whenever the ``.fit()`` function gets called, the Trainer will print the weights summary for the LightningModule.
 By default it only prints the top-level modules. If you want to show all submodules in your network, use the
-`'full'` option:
+`'max_depth'` option:
```
Comment (carmocca) [inline docs/source/common/debugging.rst:98]:
```suggestion
``max_depth`` option:
```

**Review by carmocca — APPROVED** [review]

**Comment by carmocca** [comment]
> under what circumstances would users want to disable the model summary? Is it critical we have this flag?

Maybe the summary is not correct for your model, maybe you are calling `fit` multiple times and you don't want it to appear repeatedly...

**Comment by daniellepintz** [comment]
> Maybe the summary is not correct for your model, maybe you are calling fit multiple times and you don't want it to appear repeatedly...

Hmm, if the summary is not correct that sounds like a bug, and we should operate under the assumption that we have no bugs.  The other case seems pretty rare, but I don't have a lot of background knowledge here so don't feel strongly

**Comment by williamFalcon** [comment]
@daniellepintz every trainer argument was added because there was overwhelming community need/request for it. our bar for adding any args for the lightning trainer is very high and arguments are added only in extreme circumstances where the community overwhelmingly has an issue or needs something.

**Comment by daniellepintz** [comment]
@williamFalcon makes sense, I agree user experience is very important and didn’t mean to suggest otherwise!
